### PR TITLE
Arbitrum support

### PR DESCRIPTION
**Pull request by rkdud007**
* Note : This PR is depend on #40 * 

### Arbitrum Sepolia
```
MESSAGES_INBOX = "0x7459455BDDC9f5F53af76eFe857B2064b230E4fe"
HEADERS_PROCESSOR = "0x55b87e56F0D12471e1F1a49b1619cCe7B665F2f6"
FACTS_REGISTRY = "0x8Bf5d96bE3B1722114192293F7f25C575B2C70e5"
```

### Sepolia
```
# -> ARBITRUM_SEPOLIA
L1_TO_ARBITRUM_MESSAGES_SENDER = "0x46A9C7155b2dd191aedc7E14B869E496c5fD0248"

# ParentHash Fetcher
ARBITRUM_PARENT_HASH_FETCHER = "0xEe22f06282174E7cbf7e7b864F5bF8E0b8eDDD62"
```

### CODE REVIEW

File: src/core/x-rollup-messaging/parent-hashes-fetchers/Arbitrum.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        // parentHash = outbox.roots(outputRoot);
+        // bytes32 item =  outbox.calculateItemHash(l2Sender, to, l2Block, l1Block, l2Timestamp, value, data);
+        // bytes32 merkleRoot = outbox.calculateMerkleRoot(proof, path, item);
+        // require(merkleRoot == outputRoot, "ERR_INVALID_OUTPUT_PROPOSAL");
+    
+        // fetchedForBlock =l2Block;
+        // parentHash = parentHash;
+        // TODO: proabably need some validation for block hash
+        (bytes32 outputRoot, uint256 l2Block) = abi.decode(ctx, (bytes32,uint256));
+        parentHash = outbox.roots(outputRoot);
```
Comment (marcellobardus) [inline src/core/x-rollup-messaging/parent-hashes-fetchers/Arbitrum.sol:34]: Missing `require(parentHash != bytes32(0))`
There is no guarantee that `l2Block` matches with the retrieved blockhash.
Ideally in ctx you would also encode the RLP block header and then decode the blockNumber from it to then.
`assert keccak(rlpHeader) = parentHash`
`assert l2BlockNumber == rlpHeader.getBlockNumber()`

Comment (rkdud007) [reply src/core/x-rollup-messaging/parent-hashes-fetchers/Arbitrum.sol]: resolve with https://github.com/HerodotusDev/herodotus-evm/pull/39/commits/7072e380994e0197d21e5e3fbc147afe001170a9

**Comment by rkdud007** [comment]
### One comment around integration test

Ideally, wanted to have a full integration test to check from `sending message` to `verify message is intended`. Tried approach using foundry's `createFork` to switch vm in one script. However to test full logic in this way there is little bit of consideration compare to our current contract implementation : 
- To make a test in this way, Somehow `MessagesInbox` this contract should expose a function to verify the messages is correct ( right now we don't have it :/ )
 - Also another consideration : Maybe need to wait for the message to be processed in L2
 
So I also give another suggestion if the modification is impossible. 

1) Just make test on `sending message` only, manually check `verify message is intended` step
- Pro : ye it's simple.. probably in this PR case we would chose it
- Con : It still requires some manual checking and trust 

2) Can have l1-l2 messaging testing framework ( NOTE : I m NOT talking abt implementing now, it's just idea -- maybe it can be useful opensource framework project )
- It would be looks like generalized [mocktimism](https://github.com/ethereum-optimism/mocktimism) (= generalized meaning enable other L2s like Arbitrum/zkSync etc as well ) 
- Pro : if there is solid full e2e testing framework,,,, especially in our case that needed to deploy & test LOTS of l1<>l2 message logic contracts ,, would be beneficial
- Con : ye it will takes some time,,, and idk depends on opinion, would consider as over-engineering?

File: script/IntegrationTest_Arbitrum.s.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        bytes memory _parentHashFetcherCtx = abi.encode(outputRoot, rlpHeader);
+
+        /// Context for L1ToArbitrumMessagesSender contract
+        uint256 l2GasLimit = 0x1;
```
Comment (marcellobardus) [inline script/IntegrationTest_Arbitrum.s.sol:41]: This is probably to low, I would put maybe 1m

File: script/IntegrationTest_Arbitrum.s.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        uint256 maxFeePerGas = 0x1;
+        uint256 maxSubmissionCost = 0x1;
+        bytes memory _xDomainMsgGasData = abi.encode(l2GasLimit, maxFeePerGas, maxSubmissionCost);
+
+        l1ToArbitrumMessagesSender.sendExactParentHashToL2(_parentHashFetcherCtx, _xDomainMsgGasData);
+    }
+
+    function testVerifyMessagesIsCorrect() public {
+        // To make a test in this way, Somehow `MessagesInbox` this contract should expose a function to verify the messages is correct ( right now we don't have it :/ )
```
Comment (marcellobardus) [inline script/IntegrationTest_Arbitrum.s.sol:50]: The Arbitrum `Inbox.sol` contract should be checked whether it emits an event or writes to it's own state when a message is received.
Then it should be asserted.

Comment (marcellobardus) [reply script/IntegrationTest_Arbitrum.s.sol:50]: So in Arbitrum it is the sequencer to `process the message` so your smart contract will be called by it.
https://github.com/OffchainLabs/token-bridge-contracts/blob/eebc6ac82e717484c838633f63fd0c7a19a4de51/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol#L227
Similarly as here we need to assert that the `msg.sender` is the sequencer.
In our test case we can use vm.prank to impersonate the sequencer.
https://book.getfoundry.sh/cheatcodes/prank

Comment (rkdud007) [reply script/IntegrationTest_Arbitrum.s.sol:55]: unfortunately, i assume https://github.com/OffchainLabs/token-bridge-contracts/blob/eebc6ac82e717484c838633f63fd0c7a19a4de51/contracts/tokenbridge/arbitrum/gateway/L2ArbitrumGateway.sol#L242 this liine is logic that minting l2 token based on L1 call, which inherit logic about withdrawal we cannot just call this method to get our pure message on it

File: script/IntegrationTest_Arbitrum.s.sol (hunk 2 above)
Comment (rkdud007) [reply script/IntegrationTest_Arbitrum.s.sol]: updated with https://github.com/HerodotusDev/herodotus-evm/pull/39/commits/092ad1caf39072ba362280fd0710e2de4767894d

**Review by fmkra — APPROVED** [review]

**Pull request merged**